k2/csrc/weights.cc

```diff
@@ -68,11 +69,83 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
   if (IsEmpty(fsa)) return;
   CheckInput(fsa, arc_weights);
+  ComputeForwardWeights();
+  ComputeBackardWeights();
+}
 
+// Mohri, M. 2002. Semiring framework and algorithms for shortest-distance
+// problems, Journal of Automata, Languages and Combinatorics 7(3): 321-350,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
   auto num_states = fsa.NumStates();
-  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  // TODO(haowen): compute `forward/backward_state_weights`, also check
-  // `IsTopSorted(fsa)`
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(forward_state_weights.get(), num_states, kDoubleNegativeInfinity);
+  std::vector<double> rd(num_states, kDoubleNegativeInfinity);
+
+  const auto &arcs = fsa.arcs;
+  forward_state_weights[0] = 0;
+  rd[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto src_weight = forward_state_weights[arc.src_state];
+    auto &dest_weight = forward_state_weights[arc.dest_state];
+
+    // will be `kNegativeInfinity` for non-accessible state
+    double r = rd[arc.src_state] + arc_weights[i];
+    // will be `dest_weight` for non-accessible state
+    double d = AddWeights(dest_weight, r);
+    if (!ApproxEqual(dest_weight, d, kShortestDelta)) {
+      dest_weight = d;
+      // do not overwrite original `rd` value of `src_state`
+      // as we may use it in the next iteration
+      if (arc.dest_state != arc.src_state)
+        rd[arc.dest_state] = AddWeights(rd[arc.dest_state], r);
+    }
+  }
 }
+void WfsaWithFbWeights::ComputeBackardWeights() {
+  auto num_states = fsa.NumStates();
+  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(backward_state_weights.get(), num_states,
+              kDoubleNegativeInfinity);
+  std::vector<double> rd(num_states, kDoubleNegativeInfinity);
+
+  const auto &arcs = fsa.arcs;
+  int32_t final_state = fsa.FinalState();
+  backward_state_weights[final_state] = 0;
+  rd[final_state] = 0;
+  for (auto i = static_cast<int32_t>(arcs.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto &src_weight = backward_state_weights[arc.src_state];
+    auto dest_weight = backward_state_weights[arc.dest_state];
+
+    double r = rd[arc.dest_state] + arc_weights[i];
+    double d = AddWeights(src_weight, r);
+    if (!ApproxEqual(src_weight, d, kShortestDelta)) {
+      src_weight = d;
+      // do not overwrite original `rd` value of `dest_state`
+      // as we may use it in the next iteration
+      if (arc.src_state != arc.dest_state)
+        rd[arc.src_state] = AddWeights(rd[arc.src_state], r);
+    }
+  }
+}
+
+inline double WfsaWithFbWeights::AddWeights(double weight1,
```

Comment:
I don't like the way this is repeatedly called.  I'd rather have the switch or if statement be outside the loop.
The loop can be substantially simplified, anyway; all that delta stuff is not needed.

Reply:
This is exactly a question that I wanted to ask: one way do do this is make `Wfsa` class be a templetate class on `kMax/LogSum-Weight`, is that OK? (Another way is make a virtual function, but I think that approach holds overhead of virtual call)